### PR TITLE
[Object_Detection] Add IOUSimilarity layer.

### PR DESCRIPTION
**Pull request by tanzhenyu**
Todos with ragged tensor support.

### CODE REVIEW

**Comment by bhack** [comment]
Will this only a layer or used for losses? See https://github.com/tensorflow/addons/blob/v0.10.0/tensorflow_addons/losses/giou_loss.py#L25-L61

**Comment by tanzhenyu** [comment]
> https://github.com/tensorflow/addons/blob/v0.10.0/tensorflow_addons/losses/giou_loss.py#L25-L61

Making it used for loss seems pretty easy -- just inherit from LossFunctionWrapper instead of Base Layer? I'm fine either way, depending on whether many people use iou directly as loss or not.

**Comment by bhack** [comment]
> Making it used for loss seems pretty easy -- just inherit from LossFunctionWrapper instead of Base Layer? I'm fine either way, depending on whether many people use iou directly as loss or not.

@gabrieldemarmiesse  what do you think?

**Comment by gabrieldemarmiesse** [comment]
I believe both have their use cases and it would be nice to have both. But this isn't blocking this pull request in any way. No need to add everything at once. We can always put it in the backlog.

**Comment by bhack** [comment]
@gabrieldemarmiesse Ok thanks in the meantime we could use the Layer from keras-cv and the loss from Tensorflow-addons.

**Comment by tanzhenyu** [comment]
> @gabrieldemarmiesse Ok thanks in the meantime we could use the Layer from keras-cv and the loss from Tensorflow-addons.

@bhack @gabrieldemarmiesse  Feel free to add comments & feedbacks or tag others that you think might be interested in this, always good to improve from the very beginning :-)

**Comment by bhack** [comment]
I addons was introduced as loss in 2019 by @fsx950223 with https://github.com/tensorflow/addons/pull/477

File: kerascv/layers/iou_similarity.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
+import tensorflow as tf
+
+
+class IOUSimilarity(tf.keras.layers.Layer):
```
Comment (fchollet) [inline kerascv/layers/iou_similarity.py:4]: Big question here is why a layer -- what workflows will it be used it? We do already have a IoU metric -- would it make sense for IoU to be a metric and a loss instead of a layer?

Comment (bhack) [reply kerascv/layers/iou_similarity.py:4]: Nice catch.

Comment (tanzhenyu) [reply kerascv/layers/iou_similarity.py:4]: The more I think about it, this is actually pretty different from what 1) giou loss in addons, 2) iou in keras/metrics is doing. 
Both 1) and 2) require y_true and y_pred, which needs to be the same shape, i.e.,:
y_pred: [n_boxes, 4]
y_true: [n_boxes, 4]
And the metrics/loss from it is [n_boxes] before reduction

However what we are having here is:
box1: [n_boxes_1, 4]
box2: [n_boxes_2, 4]
and we're computing the bipartite similarities, so the result would be [n_boxes_1, n_boxes_2]

More specifically, we will need to call this layer with `call(gt_boxes, anchors)`, where n_boxes_1 is usually 3-10 (which is also why I added ragged support because we don't know how many), but n_boxes_2 is always fixed (e.g., 8732 in SSD300). 
This is really used for different things, so I believe it should be a layer.

Comment (fchollet) [reply kerascv/layers/iou_similarity.py:4]: Do we expect the output of this layer to be fed to other layers? Or is it a final output of the model? If the latter, then maybe it should be a metric instead (to be used instead another layer, for instance).

Comment (AakashKumarNain) [reply kerascv/layers/iou_similarity.py:4]: IMO, this should be a metric

Comment (bhack) [reply kerascv/layers/iou_similarity.py:4]: I think that IOU and derivates could be still interesting also as loss:
https://arxiv.org/abs/1908.03851
https://arxiv.org/abs/1911.08287
